### ISSUE TITLE
Improving the Map Function- UDF Map Overlap etc.

### ISSUE THREAD

**Issue by CSSFrancis** [hyperspy/hyperspy#2936]
#### Functionality:

This is a result of a couple of different conversations and things I've been meaning to work on.

The way I see it there are a couple of things that could be added.

- `map_axes` function which would be more axes agnostic and might be useful for something like applying a Gaussian filter etc. This would probably need to be similar to the `map_overlap` function.
- Speed up the `map` function. I think that this one is kind of difficult to measure simply because this is so dependent on the chunking/ saving/ compression of the data. @uellue it might be good to talk about how you think this could be improved using the UDF and what might be the current bottle necks. 
- It might also be good to consider if some of the UDF functionality can be upstreamed to `dask` 

#### Context
- Speeding up the map function could be very helpful to making things faster but it is a little bit of a pain to update and change.

**Comment by uellue** [comment]
Yes, performance with HDF5 and Dask arrays in LiberTEM is a bit tricky because of those chunking and read amplification issues. We do have some optimizations that try to read data in a sensible way depending on the chunking scheme. Maybe we can do a head-to-head comparison of LiberTEM `map()`  with HyperSpy `map()` for a range of scenarios to see where are opportunities for improvement?

CC @sk1p

**Comment by francisco-dlp** [comment]
@uellue, could you comment on @CSSFrancis 3rd point?

> It might also be good to consider if some of the UDF functionality can be upstreamed to dask

Is this a possibility? If yes, obviously it'll be simpler to support UDF in hyperspy if it is available in dask. If no, why?

**Comment by uellue** [comment]
@francisco-dlp yes, the UDF interface could make sense as a part of Dask. From my gut feeling it could make more sense to start in HyperSpy for two reasons:

1. LiberTEM is targeting live processing, i.e. live data and/or live result display. I can imagine that applying analysis routines live while a microscope is acquiring or seeing processing results displayed live would be attractive features for HyperSpy users, but I am less sure if that is relevant for Dask in general. As far as I am aware, there's no concept of live data streams in Dask, that's why we even started LiberTEM.
2. LiberTEM has application and file reading code that would likely be relevant for HyperSpy, but not so much for Dask.

If one imagines an UDF interface purely for Dask arrays and lazy computations, it would maybe look different from what we currently have. That's why upstreaming parts of LiberTEM to Dask and at the same time making sure that our main target, live data processing, still works could be a larger effort.

Also, it is not that complicated to integrate LiberTEM UDFs with lazy HyperSpy signals right now: just a few zero-cost calls to set up the integration. See https://github.com/LiberTEM/LiberTEM/blob/master/examples/hyperspy-integration.ipynb, for example.   Adding a bit of boilerplate in some parts of HyperSpy would likely be much easier and faster than upstreaming parts of LiberTEM to Dask.

**Comment by francisco-dlp** [comment]
Yes, live processing is definitely something that could interest many HyperSpy users—it certainly interests me. Also, LiberTEM has indeed many features that are relevant for HyperSpy users, and I fully agree that everybody would benefit from integrating both ecosystem as tightly as possible.

I have only limited understanding of how LiberTEM works, but it seems to me that the performance improvements over dask in particular applications may be beneficial for users beyond our field, so I think that @CSSFrancis has a point when he wonders if parts of LiberTEM coulnd't be contributed to dask. It is obvious that the TEM-specific parts of LiberTEM don't belong to dask, but that shouldn't be a problem. I would be surprised if there are not people out there using dask for live processing, even if that's not a stated goal of the project. Integrating code into a larger project can be an arduous task full of compromises, but I bet that it is worth the effort. 

One practical question, is there a reason why LiberTEM is not in conda-forge?

**Comment by uellue** [comment]
> One practical question, is there a reason why LiberTEM is not in conda-forge?

Because it is extra work and so far it seems users managed with PyPI. We have an Issue for that https://github.com/LiberTEM/LiberTEM/issues/648 What would be practical benefits of being on conda-forge?

**Comment by francisco-dlp** [comment]
Half of the HyperSpy installations are from conda-forge. Therefore, we cannot depend on a package that is not in conda-forge.

Personally, I find that package management is easier and more reliable using conda/mamba. I may be wrong, but I think that ``conda`` is still the installation tool of choice for most people in the scientific python community.

**Comment by CSSFrancis** [comment]
From my understanding there are really two parts to this problem which are increasingly intertwined.  This is how to handle I-O and running a `map` like function or UDF. These can be largely separated into the Live data processing which is a much newer and nuanced problem and the post processing lazy operation which is quite mature and supported by `dask`.

**In the Purely Lazy Case**
From my experience with large datasets (and that seems to be where we are heading) most of the time things are I/O bound.  In that case of post processing `zarr` is a great drop-in replacement and should be the standard going forward. Dropping hdf5 seems to solve quite a bit of problems and integrates with `dask` exceptionally well. Long term I would be hesitant to switch all of the processing from `Dask` as for the non-live lazy case using `dask` should be the fastest and best supported. (Should... That might not always be the case).

**In the Live Processing Case**

 If LiberTEM is focused on live processing then it would be nice to think about how to bring some of hyperspy/pyxem functionality to live processing in LiberTEM. In that case it might be good to create an option to set a UDF background when using map.  

Right now the way that the `map` function in hyperspy works it forces the chunk to span the signal dimension.  It isn't the best in all cases but it is in my opinion the best general way of operating.  As a side effect this kind of pushes users to save in a similar fashion where they are compressing each signal dimension in the same chunk.  The hope here being that the `map` function works out of the box for most lazy signals.

From looking at the UDF this would be fairly easy to implement there as well.

I think a lot of this might end up falling to me as the map function is largely my code these days. I would be willing to work on this as it is particularly useful to my research. I could imagine a work flow for me which looks like this.

- Live Processing in LiberTEM (on Computer with fast SSD storage)
    - Centering a direct beam
    - Drift Correction in 4-5D STEM applications? (This one is personally something that my group is interested in) 
    - Validation of Data Quality (visualization, summing data etc.)
    - Reduction of Data Size (I would suggest supporting a file format which is compatible with `zarr` or `zspy` 
- Transferring to Hyperspy for Post Processing (On Local computer or Cluster with slow HDD and fast scratch Directory)
    - Cleaned and compressed data for easy transfer as well as I-O 
    - Fitting operations, peak finding, higher quality phase mapping etc.
    - ML Optimization etc

I think that all options should be shared between both packages with the main difference being that LiberTEM thinks of data as streams and optimizes for that where as Hyperspy deals with static data.  

This seems like a lot but ideally this should be fairly minimal coding if done right.  It also maybe worth thinking about LiberTEM beyond TEM.  Live large data streams are pretty common in science these days and a more general approach is probably the best way to future proof all of the operations.

**Comment by uellue** [comment]
Good question how the performance benefits could be translated! The speed comes from several factors:

* Working on small tiles that fit the L3 CPU cache, as opposed to large array chunks in Dask. LiberTEM doesn't compile a task tree and has very little overhead per tile compared to Dask array chunks. That can give a factor two or more on file formats that require decoding or computations that involve intermediate results.
* Reduce copies and re-use buffers as far as possible. Not sure how far Dask goes there.
* Optimized I/O routines. Last time I looked, Dask didn't offer much from that corner, but leaves it to users and downstream packages like HyperSpy. At least I could add memory mapping to the Dask documentation, so that part is already upstreamed in a way. ;-)
* Paying attention to data locality and avoiding to send data through a cluster where possible.
* A relatively straightforward processing flow within a UDF run and little variation in the chunking, as opposed to a task tree with free chunking. That means as long as the user-defined processing routines are fast and the data source is fast, the computation will be fast. With Dask arrays there is more flexibility, but also more ways to do things suboptimally.
* Facilities in the UDF interface that make optimizations easier, such as user-defined task data that allow re-using precomputed intermediate data or buffers.
* Some "fine details" with potentially large impact, such as setting the thread count correctly for external libraries. Otherwise libraries like OpenBLAS cause oversubscription with multiprocessing, since each process tries to use the full thread count.

Some of these can be solved by using the LiberTEM UDF interface with Dask arrays as input or output. Some can perhaps be translated to Dask as improvements. Others come from the way Dask arrays work on a fundamental level, so any lazy Dask-based processing would suffer from that, with or without LiberTEM.

**Comment by uellue** [comment]
Ah I forgot, and using multiprocessing by default. That removes the GIL from the equation and similar restrictions in the HDF5 library as an example. Even though many NumPy and external routines release the GIL, on a machine with many cores it turned out to be a bottleneck, in particular when processing smaller tiles.

**Comment by uellue** [comment]
> Half of the HyperSpy installations are from conda-forge. Therefore, we cannot depend on a package that is not in conda-forge.
> 
> Personally, I find that package management is easier and more reliable using conda/mamba. I may be wrong, but I think that `conda` is still the installation tool of choice for most people in the scientific python community.

Ok, those are good points, so I've added conda-forge support to the next release milestone. :-)

**Comment by francisco-dlp** [comment]
From your comments, it looks like we need to perform some serious benchmarking before taking a decision.

Regarding using HyperSpy features for live processing through LiberTEM, that would be cool. Using LiberTEM-live from HyperSpy has the appeal of using a syntax that people are familiar with. Imaging that the following was producing a live EELS plot from an image acquired through LiberTEM-live:

```python
live_image = libertem_stream.to_hspy_signal()
live_image.isig[:, 10:-10].sum(-1).as_signal1D(axis=1).plot()
```

**Comment by CSSFrancis** [comment]
> Ah I forgot, and using multiprocessing by default. That removes the GIL from the equation and similar restrictions in the HDF5 library as an example. Even though many NumPy and external routines release the GIL, on a machine with many cores it turned out to be a bottleneck, in particular when processing smaller tiles.

In reference to this I would go as far as to say that HDF5 is pretty much a loss for large data.  The GIL is released currently in the lazy map function (and the non lazy case as well I Optimized I/O routines. Last time I looked, Dask didn't offer much from that corner, but leaves it to users and downstream packages like HyperSpy. At least I could add memory mapping to the Dask documentation, so that part is already upstreamed in a way. ;-)believe) and uses multiprocessing by default as well just from using dask to handle the processing.   When using `zarr` most of these bottlenecks disappear and things scale much easier as well. This sees about a 10x speed improvement in general #2797.

> * Working on small tiles that fit the L3 CPU cache, as opposed to large array chunks in Dask. LiberTEM doesn't compile a task tree and has very little overhead per tile compared to Dask array chunks. That can give a factor two or more on file formats that require decoding or computations that involve intermediate results.

This makes sense in the case of binary datastreams.  In the case of compression the chunk considerations are much more involved. 

> * Reduce copies and re-use buffers as far as possible. Not sure how far Dask goes there.

IIRC Dask is pretty good at this.  Sometimes it has a problem with not releasing chunks if you start to hit the limits of the RAM but there are many tools to combat this. 

> * Optimized I/O routines. Last time I looked, Dask didn't offer much from that corner, but leaves it to users and downstream packages like HyperSpy. At least I could add memory mapping to the Dask documentation, so that part is already upstreamed in a way. ;-)

I might be a broken record but it seems like most people are moving twoards using `zarr` :).  Memmap for binary data should be the standard as well but you are correct there needs to be better documentation there. 

> * A relatively straightforward processing flow within a UDF run and little variation in the chunking, as opposed to a task tree with free chunking. That means as long as the user-defined processing routines are fast and the data source is fast, the computation will be fast. With Dask arrays there is more flexibility, but also more ways to do things suboptimally.

This is the biggest struggle I had starting with `dask` and tried to combat with the `map` function.  I think the major problem is that it just works and sometimes it is hard to know if you are actually working in an optimized way.  I think the solution here is better documentation on hyperspy's end and going from there.

The rest of the comments really go further into how `dask` schedules things which I have had little success understanding.

**Comment by magnunor** [comment]
Excellent discussion! Thanks @CSSFrancis for starting this!

Some comments:

-  I agree about `zarr`. It is so much faster! I'll probably start using it for all of my data processing.
       - However, the "folder structure" will be an issue for some people, and makes it an hassle in some situations when transferring data. For example, our webpage-filesender does not accept folders, so I would have to zip the file before sending it. And whoever receives it would have to unzip it again. Not a showstopper, but this will reduce the adoption. I understand the reason why the "folder" structure is used, but I think(?) there are some "database" ways of storing the data?
- I really like having the chunking "equal" in each dimension, like (32, 32, 32, 32) for 4D-STEM data. This so it is easy to transpose the data, ergo plotting it as a function of probe position should be as fast as plotting it as function of detector position.

-------

- I think one cause of slow processing in HyperSpy is rechunking: several functions automatically rechunk the data. For example, `rebin` and `change_dtype` will try to optimize the chunking after having rebinned the data. However, this new chunking can be really bad, causing a lot of extra data to be read from the file.

Example of this (slightly contrived example), using `zarr`:

```python
s = hs.load("test_lazy_data.zspy", lazy=True) # (510, 511, 256, 256) dataset
s1 = s.isig[1:-1]
s2 = s1.rebin(scale=(2, 1, 1, 1), rechunk=False) # Or rechunk=True (the default)
s_crop = s2.inav[:64, :64]

t0 = time.time()
#s_crop.compute(show_progressbar=False)
print(time.time() - t0)
```

`rechunk=False`: `1.2144` seconds
`rechunk=True`: `4.3897` seconds

(This is still very good compared to HDF5.)

-----

I also tested the "plotting experience" with the same dataset, with `s.change_dtype(float, rechunk=True)` vs `s.change_dtype(float, rechunk=False)`. The `rechunk=False` had much less "lag" when moving the navigator.

---------

Benchmarking: I think an important test case is the "plotting" one, where the data is accessed in an almost semi-random order. Not sure how to best benchmark this? `x = np.random.randint(0, 510)`, `y = np.random.randint(0, 511)`?

**Comment by CSSFrancis** [comment]
> * I agree about `zarr`. It is so much faster! I'll probably start using it for all of my data processing.
>       - However, the "folder structure" will be an issue for some people, and makes it an hassle in some situations when transferring data. For example, our webpage-filesender does not accept folders, so I would have to zip the file before sending it. And whoever receives it would have to unzip it again. Not a showstopper, but this will reduce the adoption. I understand the reason why the "folder" structure is used, but I think(?) there are some "database" ways of storing the data?

I think that the LMDB storage is a good place to start and was kind of hoping someone might be interested in making that part of the code more mature. 

> * I really like having the chunking "equal" in each dimension, like (32, 32, 32, 32) for 4D-STEM data. This so it is easy to transpose the data, ergo plotting it as a function of probe position should be as fast as plotting it as function of detector position.

This is one of those things that you and I have slightly different views on mostly because of the different data types we work with. In my case I rarely need to transpose the data and as a result find chunking the data in the signal dimension more efficient. It also reduces problems with the `map` function loading large amounts of data into memory.  That being said there is definitely a case to be made for equal sized chunks.   

I will also say that with equal sized chunks the `_map_iterate` function is pretty terrible and it might be worth thinking about how to implement a more chunk agnostic approach.  That gets difficult and into using something like `map_overlap` but it is something that I have been meaning to do.

**Comment by ericpre** [comment]
I never looked at it, but if I remember correctly, there is a library to deal with streaming data and which integrate with dask: https://streamz.readthedocs.io

There are some work in progress in dask, which aims at reducing some overhead:
- reducing communication between workers
- delaying materialization (or avoiding it when not necessary) of the dask graph
- optimising culling and fusion when slicing data - if I remember correctly, it is not optimised at the moment (or very little), which is bad for visualisation.

I am mentioning these because, I suspect that if we identify specific bottleneck, it may be worth fixing these in dask or providing feedback to dask developers.

Regarding some improvement in the zarr library, there should be a PR on caching decoded chunks, which we would benefit a lot in term of visualising lazy dataset. Using compression in zarr works well to balance IO with CPU.

As I am out of office, I don't have time to provide more details, but my general impression is that there are still a number of improvement possible on the dask side - or in the way we use it, which could translate to improving the dask documentation.

**Comment by francisco-dlp** [comment]
Very interesting discussion indeed. These are my conclusions so far:

* There is no such thing as a chunking strategy that is reasonable for most needs. HyperSpy currently uses chunks that contain at least one full signal, but this is sub-optimal for some common applications. This is certainly an issue, what's the best way to fix it? Better documentation is one. What about not imposing a default chunking strategy and let the user choose between some common ones?
* There is plenty of room for performance improvement in lazy HyperSpy operations:
    * Simply switching to zarr fixes many performance issues.
    * The dask community is working on fixing some overheads that impact HyperSpy's performance. Maybe the HyperSpy/LiberTEM communities can help dask pinpointing further bottlenecks and help them fix them?
    * Some HyperSpy functions (like `_map_iterate` mentioned by @CSSFrancis  above) might benefit from different implementations for different chunking styles.
* There is interest in live processing. I bet that our users will appreciate being able to perform live processing tasks with a syntax that follows HyperSpy's syntax as closely as possible. Current solutions for live processing are LiberTEM-live (based on LiberTEM and tailored for TEM) and Streamz (with dask support, not TEM-specific). There is also Nion Swift that could be used for live visualization when a GUI is required.
* There are multiple packages for big TEM data processing in Python. LiberTEM (the main topic of this issue) focuses on performance while HyperSpy (and extensions) has so far focused on using well established off-the-shelf solutions (dask). The focus on performance led to compatibility issues between LiberTEM and dask (and therefore HyperSpy). This has been fixed to a large extend in LiberTEM v0.9. Both communities should benefit from integrating their solutions (and the communities themselves) as closely as possible.

**Comment by CSSFrancis** [comment]
> * There is no such thing as a chunking strategy that is reasonable for most needs. HyperSpy currently uses chunks that contain at least one full signal, but this is sub-optimal for some common applications. This is certainly an issue, what's the best way to fix it? Better documentation is one. What about not imposing a default chunking strategy and let the user choose between some common ones?

I would be definitely for better documentation. Part of the problem here is that it all works for small datasets, it is parallel behind the scenes and so it probably faster out of the box then most of what they might be using. Most users will never think twice about how things are chunked or even what that means. I don't think that is a terrible thing but should be a consideration.

As an aside it might be good to make a distinction between file-chunks (when compression is used) and dask-chunks. 

That being said I would like to suggest something like 4 common file-chunks:

- Signal Chunked (The compression spans the signal dimension)
- Navigation Chunked (The compression spans the navigation dimension)
- Equal Chunking (Compression equally in signal and navigation)
- No Compression (Memmaped data and little cost for re chunking)

We could even implement this in a general way so people could define their default saving preference or implement a custom chunking scheme.

**Comment by francisco-dlp** [comment]
Sounds good. Just to make sure that there is no misunderstanding, could you provide an example for each of the chunking styles that you propose?

The chunking style could be stored in ``.metadata`` as the default for the signal for saving, rechunking... We could define need by adding extra keywords to ``as_lazy`` etc.

**Comment by uellue** [comment]
Regarding chunking, in https://github.com/hyperspy/hyperspy/issues/2784 are some tests of "lazy slices" as a way to address the chunking issues at the root. It looks quite promising, "just" needs support to not only subslice, but also stack such lazy slices: 

A quick summary: Conventional readers eagerly read a whole Dask chunk and pass the data to Dask as a NumPy array. That means when a subset of a Dask array is fetched, all touched chunks are read completely, and then Dask performs the slicing where larger parts may be thrown away. In the link above, a "lazy NumPy slice" is introduced. It acts as a proxy for a NumPy-like array specifically for slicing operations. It allows calculating a "slice of a sclice of a slice..." lazily, i.e. it just returns a new proxy object without touching or returning the data. The data is only fetched as soon as an operation is requested that is not implemented by the proxy.

Dask arrays can be built from such lazy slices instead of eager readers. Instead of reading data first and slicing after, the data is only read after Dask sliced the array chunks and starts to do numerical operations or returns the data to the user. That limits read operations to exactly the part that is needed. In the discussion linked above, this is shown to be particularly affective for HDF5 and other formats that implement their own chunking and use compression in combination with opertions like frame picking that only need a small subset of the data.

It lifts several of the current constraints:

1. Dask chunks can be very large without causing read amplification when small subsets are accessed. That allows working at an ideal size for throughput, between 100 and 500 MB for Dask.
2. The Dask chunking scheme has way less impact on the characteristics of operations that access subsets since only the required subsets of a Dask chunk are actually read.
3. The chunking of the file can be very small. About 1 MB is optimal for compressed HDF5. That reduces issues of read amplification on the file format side.

To be fully production-ready, the lazy slices still have to support stacking so that rechunking and reshaping are efficient operations, too. In LiberTEM we also have code that can read small subsets of a file efficiently, which could be used to implement the back-end for such lazy slices. And, for LiberTEM such lazy slices would be an ideal abstraction layer for our own reader code. That means, in effect, we could share all I/O code and have a very good generic reader for very large data. It presents a NumPy array interface similar to HDF5, just for arbitrary file types.

**Comment by uellue** [comment]
> From your comments, it looks like we need to perform some serious benchmarking before taking a decision.
> 
> Regarding using HyperSpy features for live processing through LiberTEM, that would be cool. Using LiberTEM-live from HyperSpy has the appeal of using a syntax that people are familiar with. Imaging that the following was producing a live EELS plot from an image acquired through LiberTEM-live:
> 
> ```python
> live_image = libertem_stream.to_hspy_signal()
> live_image.isig[:, 10:-10].sum(-1).as_signal1D(axis=1).plot()
> ```

That looks pretty cool indeed. :-) The operation you show is very simple to implement with LiberTEM and use for live plotting: https://gist.github.com/uellue/f964c9f158378f595778ec254d7f160c

Probably one could discuss how to make a more "HyperSpy-like" interface?
